labrad/decorators.py

```diff
@@ -21,202 +21,140 @@
+        # 3) The empty tag '' is only allowd on the first argument, and only
+        #    if all other arguments are optional.
+        # 4) If more than one argument are required, we expect to always receive
+        #    a tuple and unpack it
+        # 5) If only a single argument is allowed, we never unpack tuples
+        # 6) If both =1 and >1 arguments are allowed, it is ambiguous whether to
+        #    unpack tuples, so this case is not allowed:  The first argument
+        #    cannot be a tuple or '?' tag if the second argument is optional.
+        
+        argspec = inspect.getargspec(self.func)
+        args = argspec.args[2:]
+
+        if inspect.isgeneratorfunction(func):
+            self.func = defer.inlineCallbacks(func)
+
+        for arg in args:
+            if arg not in params:
+                params[arg] = ['?']
+
+        for p in params.keys():
             if p not in args:
-                raise Exception("'%s' is not a valid parameter." % p)
-            # turn single string annotations into lists
-            if isinstance(params[p], str):
+                raise ValueError("Setting parameter {} not accepted by function".format(p))
+            if isinstance(params[p], basestring):
                 params[p] = [params[p]]
+            #params[p] = [T.parseTypeTag(tag) for tag in params[p]]
+
 
         Nparams = len(args)
-        Noptional = 0 if defaults is None else len(defaults)
+        Noptional = len(argspec.defaults) if argspec.defaults else 0
         Nrequired = Nparams - Noptional
 
-        if Nparams == 0:
-            accepts_s = [''] # only accept notifier
-            accepts_t = [T.parseTypeTag(s) for s in accepts_s]
-
-            @wraps(f)
-            def handleRequest(self, c, data):
-                return f(self, c)
-
-        elif Nparams == 1:
-            accepts_s = params.get(args[0], [])
-            accepts_t = [T.parseTypeTag(s) for s in accepts_s]
-
-            if Nrequired == 0:
-                # if accepted types were specified, add '' to the list
-                # we don't add '' if the list of accepted types is empty,
-                # since this would make '' the ONLY accepted type
-                if len(accepts_t) and T.LRNone() not in accepts_t:
-                    accepts_s.append(': defaults [%s=%r]' \
-                                     % (args[0], defaults[0]))
-                    accepts_t.append(T.LRNone())
-
-                @wraps(f)
-                def handleRequest(self, c, data):
-                    if data is None:
-                        return f(self, c)
-                    return f(self, c, data)
-
+        if Nrequired > 1:
+            self.unpack = "always"
+        elif Nparams > 1:
+            self.unpack = "optional"
+        else: # one or fewer arguments
+            self.unpack = "never"
+
+        self.allow_none = Nrequired == 0
+
+        if Nparams:
+            for tag_str in params[args[0]]:
```

Comment:
I've tried to be consistent elsewhere in calling the string form a "tag" and the parsed object a "type". of course, the name "type" is a name of a builtin class, so usually "t" is used for that. It'd be nice to stay consistent with that convention. (One place where I know we're inconsistent is in the FlatData object which has a "tag" field that contains the parsed type object; we might want to fix that, too.)